Deprecate client module dependencies (#25250)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

The embedded Kotlin has been updated to link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.21[Kotlin 1.8.21].

=== Deprecations

==== Deprecated `ClientModule` dependencies

> `ClientModule` dependencies are deprecated and will be removed in Gradle 9.0.

Author: The `DependencyHandler` class-level javadoc does explain these though. We can link to the class-level javadoc, or we can write new gradle docs for it

> `ClientModule` dependencies are deprecated and will be removed in Gradle 9.0.

Reviewer: I'll add a reminder for myself to add this to the docs (in the future). For now we can link to javadocs.

> `ClientModule` dependencies are deprecated and will be removed in Gradle 9.0.

Author: Sounds good. In terms of priority, it _is_ getting removed in 9.0, so I'm not sure how important it is. For reference though, here are some original docs which explain the feature. They were removed at some point...

https://docs.gradle.org/4.4/userguide/dependency_management.html#sub:client_module_dependencies

Client module dependencies were originally intended to allow builds to override incorrect or
missing component metadata of external dependencies by defining the metadata locally. This

> functionality has since been replaced by <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>

Reviewer: ```suggestion
functionality has since been replaced by <<component_metadata_rules.adoc#sec:component_metadata_rules,Component Metadata Rules>>.
```

Consider the following client module dependency example:

=====
[.multi-language-sample]
======
.build.gradle.kts
[source,kotlin]
----
dependencies {
    implementation(module("org:foo:1.0") {
        dependency("org:bar:1.0")
    })
}
----
======
[.multi-language-sample]
======
.build.gradle
[source,groovy]
----
dependencies {
    implementation module("org:foo:1.0") {
        dependency "org:bar:1.0"
    }
}
----
======
=====

This can be replaced with the following component metadata rule:

=====
[.multi-language-sample]
======
.build.gradle.kts
[source,kotlin]
----
dependencies {
    components {
        withModule("org:foo") {
            listOf("compile", "runtime").forEach { base ->

>                 withVariant(base) {

Reviewer 2: Would allVariants be appropriate instead of listing variant names?

                    withDependencies {
                        add("org:bar:1.0")
                    }
                }
            }
        }
    }

    implementation("org:foo:1.0")
}
----
======
[.multi-language-sample]
======
.build.gradle
[source,groovy]
----
dependencies {
    components {
        withModule("org:foo") {
            ["compile", "runtime"].each { base ->
                withVariant(base) {
                    withDependencies {
                        add("org:bar:1.0")
                    }
                }
            }
        }
    }

    implementation "org:foo:1.0"
}
----
======
=====

> While the updated solution may be more verbose, we decided that

Reviewer: ```suggestion
While the updated solution may be more verbose, 
```

> reducing the number of ways to achieve the same result is worth

Reviewer: ```suggestion
reducing the number of ways to achieve the same result is a worthwhile
```

> trade-off. Moving forward, component metadata rules

Reviewer 2: Should we include an example that uses a metadata rule class?

> the trade-off. In any case, component metadata rules like this

Reviewer: ```suggestion
trade-off. Moving forward, component metadata rules
```

> should live within build-logic and thus should not clutter up

Reviewer: ```suggestion
should live within build-logic and should not clutter up
```

The embedded Kotlin has been updated to link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.21[Kotlin 1.8.21].

=== Deprecations

==== Deprecated `ClientModule` dependencies

`link:{javadocPath}/org/gradle/api/artifacts/ClientModule.html[ClientModule]` dependencies are deprecated and will be removed in Gradle 9.0.

Client module dependencies were originally intended to allow builds to override incorrect or

> missing component metadata of external dependencies by defining the metadata locally. This

Reviewer 3: 🧐 Use one sentence per line in asciidoc
